azureDatabaseUtils/src/main/java/bio/terra/workspace/azureDatabaseUtils/database/DatabaseService.java

```diff
@@ -80,4 +89,97 @@ public void restoreNamespaceRoleAccess(String namespaceRole) {
 
     databaseDao.restoreLoginPrivileges(namespaceRole);
   }
+
+  public void pgDump(
+      String sourceDbName,
+      String sourceDbHost,
+      String sourceDbPort,
+      String sourceDbUser,
+      String pgDumpFilename,
+      String destinationWorkspaceId,
+      String blobstorageDetails) {
+    logger.info("running DatabaseService.pgDump against {}", sourceDbName);
+    logger.info("destinationWorkspaceId: {}", destinationWorkspaceId);
+    try {
+      // Grant the database role (sourceDbName) to the workspace identity (sourceDbUser).
+      // In theory, we should be revoking this role after the operation is complete.
+      // We are choosing to *not* revoke this role for now, because:
+      // (1) we could run into concurrency issues if multiple users attempt to clone the same
+      // workspace at once;
+      // (2) the workspace identity can grant itself access at any time, so revoking the role
+      // doesn't protect us.
+      databaseDao.grantRole(sourceDbUser, sourceDbName);
+
+      List<String> commandList =
+          generateCommandList("pg_dump", sourceDbName, sourceDbHost, sourceDbPort, sourceDbUser);
+      Map<String, String> envVars = Map.of("PGPASSWORD", determinePassword());
+      LocalProcessLauncher localProcessLauncher = new LocalProcessLauncher();
+      localProcessLauncher.launchProcess(commandList, envVars);
+
+      storage.streamOutputToBlobStorage(
+          localProcessLauncher.getInputStream(),
+          pgDumpFilename,
+          destinationWorkspaceId,
+          blobstorageDetails);
+
+      String output = checkForError(localProcessLauncher);
+      logger.info("pg_dump output: {}", output);
+
+    } catch (PSQLException ex) {
+      logger.error("process error: {}", ex.getMessage());
```

Comment:
why ignore this exception?

Reply:
good question. I have to admit that I lifted this part from WDS's code, and didn't question it - I'll talk to someone on that team and see if there's some non-obvious reason for ignoring it

Reply:
I got an answer: they implemented it this way because, for WDS, this process is part of app startup, and they wanted it to continue even if some part of the dump/restore failed, so that they could gracefully fall back to an empty database.

I think that reasoning shouldn't apply in our case, because (as part of our Leonardo changes) we're going to have apps use databases *if they exist* - and if they don't exist, create them from scratch. 

Something I'm not sure about: if this process throws an exception, and the flight step fails as a result, does that cancel the whole flight? or can the flight continue to the next step?